jumpscale/sals/process/__init__.py

```diff
@@ -84,595 +141,833 @@ def execute(
 
 
 def is_alive(pid):
-    """Checks if pid is Running
+    """Check whether the given PID exists in the current process list.
 
-    Arguments:
-        pid (int) -- pid of the process to be checked
+    Args:
+        pid (int): Process ID (PID) to be checked.
 
     Returns:
-        [bool] -- True if process is running
+        bool: True if the given PID exists in the current process list, False otherwise.
     """
-    pid = int(pid)
     return psutil.pid_exists(pid)
 
 
 def is_installed(cmd):
-    """[summary]
-    Checks if a specific command is available on system e.g. curl
-    Arguments:
-        cmd {str} -- command to be checked
+    """Checks if a specific command is available on system e.g. curl.
+
+    Args:
+        cmd (str): Command to be checked.
 
     Returns:
-        [bool] -- True if command is installed
+        bool: True if command is available, False otherwise.
     """
-    rc, _, _ = execute("which %s" % cmd, die=False)
-    if rc:
-        return False
-    else:
-        return True
-
+    rc, _, _ = execute(f"which {cmd}", die=False)
+    return rc == 0
 
-def kill(pid, sig=signal.SIGTERM.value):
-    """Kill a process with a signal
 
-    Arguments:
-        pid (int) -- pid of the process to be killed
+def kill(proc, sig=signal.SIGTERM, timeout=5, sure_kill=False):
+    """Kill a process with a specified signal.
 
-    Keyword Arguments:
-        sig {int]} -- which signal you want to kill the process with (default: {signal.SIGTERM.value})
+    Args:
+        proc (int or psutil.Process): Target process ID (PID) or psutil.Process object.
+        sig (signal, optional): See signal module constants. Defaults to signal.SIGTERM.
+        timeout (int, optional): How long to wait for a process to terminate (seconds) before raise exception
+            or, if sure_kill=True, send a SIGKILL. Defaults to 5.
+        sure_kill (bool, optional): Whether to fallback to SIGKILL if the timeout exceeded for the terminate operation or not. Defaults to False.
 
     Raises:
-        j.exceptions.RuntimeError: in case killing process failed
+        j.exceptions.Runtime: In case killing the process failed.
+        j.exceptions.Permission: In case the permission to perform this action is denied.
 
     Returns:
-        [type] -- [description]
+        None
     """
-    pid = int(pid)
-    sig = int(sig)
-    proc = psutil.Process(pid)
     try:
+        if isinstance(proc, int):
+            proc = get_process_object(proc)
+        if proc.status() == psutil.STATUS_ZOMBIE:
+            return True
         proc.send_signal(sig)
+        # Wait for a process to terminate
+        # If PID no longer exists return None immediately
+        # If timeout exceeded and the process is still alive raise TimeoutExpired exception
+        proc.wait(timeout=timeout)
+        j.logger.debug(f"the process with PID: {proc.pid} was terminated with sig: {sig}.")
+        return True  # XXX only to pass current tests
```

Comment:
The same as previous comments, we can remove the return and change the tests as long as there's no caller which depends on this return value (Also this can be updated too).

Reply:
The same applies for the rest of the function

Reply:
> The same as previous comments, we can remove the return and change the tests as long as there's no caller which depends on this return value (Also this can be updated too).

It the same as the previous replay.

Reply:
> The same applies for the rest of the function

It the same as the previous replay.